src/index.ts

```diff
@@ -0,0 +1,449 @@
+import * as crypto from 'crypto'
+import * as ethUtil from 'ethereumjs-util'
+
+const bs58check = require('bs58check')
+const randomBytes = require('randombytes')
+const scryptsy = require('scrypt.js')
+const uuidv4 = require('uuid/v4')
+
+// parameters for the toV3() method
+
+interface V3Params {
+  kdf: string
+  cipher: string
+  salt: Buffer
+  iv: Buffer
+  uuid: Buffer
+  dklen: number
+  c: number
+  n: number
+  r: number
+  p: number
+}
+
+function mergeToV3ParamsWithDefaults(params?: Partial<V3Params>): V3Params {
+  const v3Defaults: V3Params = {
+    cipher: 'aes-128-ctr',
+    kdf: 'scrypt',
+    salt: randomBytes(32),
+    iv: randomBytes(16),
+    uuid: randomBytes(16),
+    dklen: 32,
+    c: 262144,
+    n: 262144,
+    r: 8,
+    p: 1,
+  }
+
+  if (!params) {
+    return v3Defaults
+  }
+  return {
+    cipher: params.cipher || 'aes-128-ctr',
+    kdf: params.kdf || 'scrypt',
+    salt: params.salt || randomBytes(32),
+    iv: params.iv || randomBytes(16),
+    uuid: params.uuid || randomBytes(16),
+    dklen: params.dklen || 32,
+    c: params.c || 262144,
+    n: params.n || 262144,
+    r: params.r || 8,
+    p: params.p || 1,
+  }
+}
+
+interface KDFParams {
+  c: number
+  prf: string
+  dklen: number
+  n: number
+  r: number
+  p: number
+  salt: string
+}
+
+/**
+ * Based on the parameter list passed to the Wallet.prototype.toV3() method this
+ * returns a list of parameters for running the key derivation function.
+ * @param params params passed into the .toV3() method
+ */
+function mergeKDFParamsWithDefaults(params: V3Params): KDFParams {
+  const kdfDefaults = {
+    c: 262144,
+    prf: 'hmac-sha256',
+    n: 262144,
+    r: 8,
+    p: 1,
+    salt: params.salt.toString('hex'),
+  }
+
+  return {
+    dklen: params.dklen,
+    salt: kdfDefaults.salt,
+    c: params.c || kdfDefaults.c,
+    prf: kdfDefaults.prf,
+    n: params.n || kdfDefaults.n,
+    r: params.r || kdfDefaults.r,
+    p: params.p || kdfDefaults.c,
+  }
+}
+
+function stripUnusedKDFParamsForPBKDF2(params: KDFParams): Partial<KDFParams> {
```

Comment:
Is there a particular reason why `KDFParams` is handled like this? One alternative I see is to have a specific interface for each kdf, and a general type which is their union.

Reply:
Yes that's a much better idea - will go with that